File: behaviorial/visitor/python/concrete_element.py


from xml.dom.minidom import Document

from visitor import DocumentVisitor


class Report(Document):
    def generate_content(self):
        print("Generating report...")

    def accept(self, visitor: DocumentVisitor):
        visitor.visit_report(self)


class Invoice(Document):
    def generate_content(self):
        print("Creating invoice...")

    def accept(self, visitor: DocumentVisitor):
        visitor.visit_invoice(self)


class Resume(Document):
    def generate_content(self):
        print("Preparing resume...")

    def accept(self, visitor: DocumentVisitor):
        visitor.visit_resume(self)

File: behaviorial/visitor/typescript/concrete-element.ts
import { DocumentVisitor } from './visitor';
import { Document } from './element';

export class Report implements Document {
	accept(visitor: DocumentVisitor): void {
		visitor.visitReport(this);
	}

	generateReport(): void {
		console.log('Generating report...');
	}
}

export class Invoice implements Document {
	accept(visitor: DocumentVisitor): void {
		visitor.visitInvoice(this);
	}

	createInvoice(): void {
		console.log('Creating invoice...');
	}
}

export class Resume implements Document {
	accept(visitor: DocumentVisitor): void {
		visitor.visitResume(this);
	}

	prepareResume(): void {
		console.log('Preparing resume...');
	}
}
